plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -79,97 +82,124 @@ Then, in the _body_, we track the term and type application (`Apply` or `TyInst`
 computeArity = \case
-    LamAbs _ _ _ body -> MkTerm : computeArity body
-    TyAbs _ _ _ body  -> MkType : computeArity body
+    LamAbs _ _ _ body -> (first ((:) MkTerm) (computeArity body))
+    TyAbs _ _ _ body  -> (first ((:) MkType) (computeArity body))
     -- Whenever we encounter a body that is not a lambda or type abstraction, we are done counting
-    _                 -> []
+    tm                -> ([],tm)
+
+-- | A term or type argument.
+data Args tyname name uni fun ann =
+  MkTermArg (Term tyname name uni fun ann)
+  | MkTypeArg (Type tyname uni ann)
+
+-- | A list of type or term argument(s) that are being applied.
+type ArgOrder tyname name uni fun ann = [Args tyname name uni fun ann]
+
+-- | A pair of argument and the annotation of the term being applied to,
+-- so a term that was traversed can be built back with `mkApps`.
+type ArgOrderWithAnn tyname name uni fun ann =
+  [(Args tyname name uni fun ann, ann)]
+
+-- | Takes a term or type application expression and returns the function
+-- being applied and the arguments to which it is applied
+collectArgs :: Term tyname name uni fun ann
+  -> (Term tyname name uni fun ann, ArgOrderWithAnn tyname name uni fun ann)
+collectArgs expr
+  = go expr []
+  where
+    go (Apply ann f a) as      = go f ((MkTermArg a, ann):as)
+    go (TyInst ann f tyArg) as = go f ((MkTypeArg tyArg, ann):as)
+    go e as                    = (e, as)
+
+-- | Apply a list of term and type arguments to a function in potentially a nested fashion.
+mkApps :: Term tyname name uni fun ann
+  -> ArgOrderWithAnn tyname name uni fun ann
+  -> Term tyname name uni fun ann
+mkApps f ((MkTermArg tmArg, ann) : args) = mkApps (Apply ann f tmArg) args
+mkApps f ((MkTypeArg tyArg, ann) : args) = mkApps (TyInst ann f tyArg) args
+mkApps f []                              = f
+
+-- | Given the arity of a function, and the list of arguments applied to it, return whether it is
+-- fully applied or not.
+isFullyApplied :: Arity -> ArgOrder tyname name uni fun ann -> Bool
+isFullyApplied [] (_argsOrder:_as) = True -- over-application
+isFullyApplied (_arity:_) [] = False -- under-application
+isFullyApplied [] [] = True
+isFullyApplied lamOrder argsOrder =
+  -- start comparing from the end because there may be over-application
+  case (last lamOrder, last argsOrder) of
+    (MkTerm, MkTermArg _) -> isFullyApplied (init lamOrder) (init argsOrder)
+    (MkType, MkTypeArg _) -> isFullyApplied (init lamOrder) (init argsOrder)
+    _                     -> False
+
+-- | Inline fully applied functions iff the body of the function is `acceptable`.
+inlineSat :: forall tyname name uni fun ann. InliningConstraints tyname name uni fun
+    => Term tyname name uni fun ann -- ^ The `body` of the `Let` term.
+    -> InlineM tyname name uni fun ann (Term tyname name uni fun ann)
+inlineSat appOrTyInstTm =
+    case appOrTyInstTm of
+      -- If the term is a term or type application, check it's applying to a var that we may inline
```

Comment:
why do we need this check? Shouldn't everything work gracefully in the case of 0 arguments?

Reply:
yes, it should indeed.